### ISSUE TITLE
Don't lazy initialize PS decoder

### ISSUE THREAD

**Issue by watsonbox** [watsonbox/pocketsphinx-ruby#12]
As pointed out in #10, initializing the Pocketsphinx decoder actually changes configuration, which can be confusing since these values appear to be updated after the first decoding operation.

**Comment by nshmyrev** [comment]
Looks cool, thank you!

**Comment by ojak** [comment]
I'm still seeing that `Decoder.new` is overwriting explicit configuration flags.  Here's what I'm seeing after the above-listed patch:
```ruby
> configuration = Pocketsphinx::Configuration.default
> configuration['cmninit']
=> "8.0"
> decoder = Pocketsphinx::Decoder.new(configuration)
> decoder.configuration['cmninit']
=> "40,3,-1"
```
I'm thinking that `Configuration.default` should essentially be initializing the same way as `Decoder.new`, and `Decoder` should only fill in flags that are not already set.

`Decoder.new()` and `Decoder.new(Configuration.default)` should result in identical decoders.

**Comment by watsonbox** [comment]
@ojak Yes as I said in #10: "Note that there would still be a 'two-phase' initialization - one when creating a `Configuration` and one when creating a `Decoder`."  In Pocketsphinx terms, the `Configuration` represents the command line options (or user options) which are passed to the decoder.  Then `ps_init` initializes a decoder which is represented by `Decoder`.

This gem is intended to wrap Pocketsphinx and while I'm not against adding a layer of useful behavior where absolutely necessary (i.e. #10), I don't want it to become a replacement implementation for various parts of Pocketsphinx.  In this case, digging around doing custom initialization with the `feat.params` feels like precisely that (i.e. providing a duplicate implementation of the private [`ps_init_defaults`](https://github.com/cmusphinx/pocketsphinx/blob/master/src/libpocketsphinx/pocketsphinx.c#L109)).

It looks like in theory this could be resolved quite easily in Pocketsphinx, by adding (or moving) the call to [`ps_init_defaults`](https://github.com/cmusphinx/pocketsphinx/blob/master/src/libpocketsphinx/pocketsphinx.c#L109) to the end of [`ps_default_search_args`](https://github.com/cmusphinx/pocketsphinx/blob/master/src/libpocketsphinx/pocketsphinx.c#L163).  Or even by exposing `ps_init_defaults` in the public API, with the configuration as a parameter instead of the decoder.  That's something that would need to be raised in the Pocketsphinx forums.

I'll add a note to the Decoder section of the README to avoid confusion as much as possible for now.

**Comment by ojak** [comment]
@watsonbox Ok.  If that's the case, passing a configuration directly into `Decoder.new` will never work.  Why not just remove the `configuration` parameter from the initializer entirely and replace it with `Pocketsphinx::Configuration.default`?

That would at least help further reduce confusion, since the `Decoder.new` really doesn't obey the configuration parameter since it's always going to be overwritten anyhow?

**Comment by watsonbox** [comment]
Not sure what you mean when you say passing a configuration will never work.  There are different configurations described in the documentation for [keyword spotting](https://github.com/watsonbox/pocketsphinx-ruby#keyword-spotting) and using [grammars](https://github.com/watsonbox/pocketsphinx-ruby#grammars).  These are just two possibilities that I've singled out for special attention but there could be any number of reasons why a user might wish to adjust the decoder configuration.

The decoder only overwrites a small subset of configuration settings relating to the acoustic model.

**Comment by ojak** [comment]
Sorry.  To clarify, what I mean is that the small subset of overwritten parameters you're referring to are absolutely critical to providing an accurate decode, and without them, the decoder fails to accurately decode.

I agree with you with regards to not wanting to duplicate the configuration of the underlying Pocketsphinx library's own local configuration, which is totally fine.  But the setup here results in a silent replacement of wrapper configuration variables that most developer will reasonably expect are reliably configuring the tool.  Since the configuration problem/bug is effectively being hidden during the wrapper communication with the library, this will pretty much guarantee confusion.  I'm just suggesting some sort of fix based on my own lengthy experience of trying to get the gem to work (and nearly aborting using the gem at all).

It's an amazing and powerful gem, and I'm really just trying to help figure out a way to make it more developer-friendly and accessible.

**Comment by watsonbox** [comment]
Okay no worries - I propose to close this issue off and either continue discussion on your PR, or on a new issue/PR related to any other concrete proposals for improvement.

This issue was really just to ensure that all config init was complete before starting the first decode so as to align the gem's behavior with that of Pocketsphinx itself.

The 'silent' replacement of wrapper variables based on the acoustic model is a Pocketsphinx feature, analogous to the same replacement of command line arguments when using it on the command line. For now at least, that's just how it works.

**Comment by ojak** [comment]
Ok.  Is there a running TODO list for this gem?  This would probably be a good enhancement to add to that list.

**Comment by watsonbox** [comment]
I've been using issues and PRs for that, as long as the proposal is relatively concrete.  That way it's easier to discuss, and things don't simply sit on the list forever.
